### ISSUE TITLE
GNU sparse tar file not properly handled

### ISSUE THREAD

**Issue by nitmir** [mxmlnkn/ratarmount#23]
Hi
GNU sparse tar file (https://www.gnu.org/software/tar/manual/html_node/sparse.html) are not properly handled.

Here a sample tar file following the PAX Format version 1.0 format (https://www.gnu.org/software/tar/manual/html_node/Sparse-Formats.html) generated using GNU tar 1.30 with the command: ```tar --format=pax -cSzf ../sample1.tar.gz *```

[sample1.tar.gz](https://github.com/mxmlnkn/ratarmount/files/4214466/sample1.tar.gz)

The file contains 3 identical files, two of them being sparse files. The file "01.sparse1.bin" was created 
as a 10MB file filled with zeros and with the word "toto" writed at a random offset in the file. the files "02.normal1.bin" is a copy of this file. The file "01.sparse1.bin" is transformed into sparse files using ```fallocate -d 01.sparse1.bin``` and then copied into "03.sparse1.bin"

```
$ ls -alnh *
-rw-r--r-- 1 1000 1000 11M févr. 17 15:54 01.sparse1.bin
-rw-r--r-- 1 1000 1000 11M févr. 17 15:54 02.normal1.bin
-rw-r--r-- 1 1000 1000 11M févr. 17 15:53 03.sparse1.bin
$ du -sh *
8,0K	01.sparse1.bin
11M	02.normal1.bin
8,0K	03.sparse1.bin
$ md5sum *
832c78afcb9832e1a21c18212fc6c38b  01.sparse1.bin
832c78afcb9832e1a21c18212fc6c38b  02.normal1.bin
832c78afcb9832e1a21c18212fc6c38b  03.sparse1.bin
```

If I mount this tar file using ratarmount the sparse files are corrupted
```
$ ratarmount -c sample1.tar.gz mount
Creating offset dictionary for /tmp/ratarmount/sample1.tar.gz ...
Creating new SQLite index database at /tmp/ratarmount/sample1.tar.gz.index.sqlite
Creating offset dictionary for /tmp/ratarmount/sample1.tar.gz took 0.10s
$ cd mount/
$ md5sum *
01fcca0bffe4125261d2e2a85fa8223b  01.sparse1.bin
832c78afcb9832e1a21c18212fc6c38b  02.normal1.bin
390d8adb1cf6f6c1ef76345fc9758dcc  03.sparse1.bin
```

Even if ratarmount do not handle sparse file, the mounted sparse files are still corrupted compared to a tar implementation not supporting sparse files. 01.sparse1.bin and 03.sparse1.bin should be identical because there are identical inside the tar file.
Cf https://www.gnu.org/software/tar/manual/html_node/PAX-1.html#SEC193
"The format is designed in such a way that non-posix aware tars and tars not supporting GNU.sparse.* keywords will extract each sparse file in its condensed form with the file map prepended and will place it into a separate directory. Then, using a simple program it would be possible to expand the file to its original form even without GNU tar."

I use a lot of big sparse tar and I stumbled on ratarmount which seems a lot more faster than archivemount. I would love to be able to use ratarmount with this files.

**Comment by nitmir** [comment]
It seems to me that ratarmount correctly skip the sparse header but do not expand the holes, hence the end a the sparse file is either garbage from the next headers and files in the tar archive, either truncated because we reach EOF.

**Comment by mxmlnkn** [comment]
I see two parts of this issue:

 1. Bug: sparse files contain wrong data (caused by wrong size) 
    The bug stems from tarfile reporting the expanded size and there seems to be no public API way to query for the sparse file. tarfile even reports the PAX headers but they contain only the expanded real size, too. So there seems to be no easy workaround. One option would be to manually parse the tar file info metadata or search in the tarfile for some internal information about the compressed size or implement the sparse reading feature, see below. All but the last are kinda ugly.

 2. Feature: sparse files are not expanded. 
   I could somehow use tarfile to skip to that file header inside the tar, read it again and then let it do the interpretation for us. I'll have to look into why I didn't do this (is it possible without reading the whole beginning of the tar again. Can I seek inside the file,...). Alternatively, I could read into and implement the file map specification myself but then I still might require the compressed size of those metadata ...

**Comment by mxmlnkn** [comment]
Ok, I got an idea and the weekend is here. I quickly tried seeking to the tar offset for the file while iterating over it with tarfile, however, that did not work for some reason. The idea would now be to still use tarfile as another abstraction layer to read the file contents (at least for sparse files, maybe I'll keep reading the raw data for normal files for performance reasons). The way I can get tarfile to directly seek to the file in question would be by giving it an io buffered file object which hides the unneeded parts of the original file. The layers for access to the actual data are beginning to get insane with posix, fuse, bz2reader, three or more layers in python, ... but oh well. :)

**Comment by mxmlnkn** [comment]
Ok, I pushed a draft to develop. Maybe you can test it out. You will have to use `-c` to force the index to be created from scratch.

I also added your tar into the tests folder because they looked like minimal examples without any personal information. If you are opposed to this, please leave a note!

What's missing now is backwards compatibility. I only want to give a warning if the index is too old and not force an exception. However, currently the additional columns are a problem at some parts.

**Comment by nitmir** [comment]
Hi sorry for the late response.

You're welcome to use the tar file in the tests.

The commit seems to fix the issue.
Unfortunately I also found bugs related to sparse files directly in the python stdlib module tarfile and those are impacting ratarmount.

Theses bugs only concerns sparse files where actual data is bigger than 8Go (0o77777777777 bytes).

I reported one of them here: https://bugs.python.org/issue39688

The other one I have found is about the _apply_pax_info in https://github.com/python/cpython/blob/master/Lib/tarfile.py#L1341. The value used for self.size depends on the order of the tar pax header attributes: for a big sparse file, it is possible to have the GNU.sparse.realsize (size of the file with the holes expanded) AND size (size of the tar data block) header. In this function the GNU.sparse.realsize should be always used in preference of the size header then available
